File: packages/python/src/ghagen/lint/output.py
"""Output formatters for lint violations.

Three formats are supported:

- ``format_human``   — pretty terminal output for humans
- ``format_json``    — machine-readable JSON for CI and tooling
- ``format_github``  — GitHub Actions workflow-command annotations
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from ghagen.lint.violation import Severity, Violation


def _counts(violations: Iterable[Violation]) -> tuple[int, int]:
    errors = sum(1 for v in violations if v.severity == Severity.ERROR)
    warnings = sum(1 for v in violations if v.severity == Severity.WARNING)
    return errors, warnings


# ---------------------------------------------------------------- human


def format_human(violations: list[Violation]) -> str:
    """Render violations as human-readable terminal output."""
    if not violations:
        return "No violations found.\n"

    lines: list[str] = []
    for v in violations:
        loc = v.location
        if loc.file is not None and loc.line is not None:
            prefix = f"{loc.file}:{loc.line}"
        elif loc.file is not None:
            prefix = str(loc.file)
        else:
            prefix = "<unknown>"

        lines.append(f"{prefix}: {v.severity}[{v.rule_id}]")
        lines.append(f"  {v.message}")
        lines.append(f"  Symbolic path: {loc.symbolic}")
        if v.hint:
            lines.append(f"  hint: {v.hint}")
        lines.append("")  # blank line between entries

    errors, warnings = _counts(violations)
    total = errors + warnings
    summary = (
        f"Found {total} violation{'s' if total != 1 else ''} "
        f"({errors} error{'s' if errors != 1 else ''}, "
        f"{warnings} warning{'s' if warnings != 1 else ''})."
    )
    lines.append(summary)
    return "\n".join(lines) + "\n"


# ---------------------------------------------------------------- json


def _violation_to_dict(v: Violation) -> dict:
    loc = v.location
    return {
        "rule_id": v.rule_id,
        "severity": v.severity.value,
        "message": v.message,
        "location": {
            "file": str(loc.file) if loc.file is not None else None,
            "line": loc.line,
            "symbolic": loc.symbolic,
        },
        "hint": v.hint,
    }


def format_json(violations: list[Violation]) -> str:
    """Render violations as a JSON document."""
    errors, warnings = _counts(violations)
    payload = {
        "violations": [_violation_to_dict(v) for v in violations],
        "summary": {"errors": errors, "warnings": warnings},
    }
    return json.dumps(payload, indent=2)


# ---------------------------------------------------------------- github


def _escape_gh_message(msg: str) -> str:
    """Escape a message for GitHub workflow-command encoding.

    See: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
    """
    return (
        msg.replace("%", "%25")
        .replace("\r", "%0D")
        .replace("\n", "%0A")
    )


def _escape_gh_property(value: str) -> str:
    """Escape a property value (file=, title=, etc.) for GH commands."""
    return (
        value.replace("%", "%25")
        .replace("\r", "%0D")
        .replace("\n", "%0A")
        .replace(":", "%3A")
        .replace(",", "%2C")
    )


def format_github(violations: list[Violation]) -> str:
    """Render violations as GitHub Actions workflow-command annotations."""
    if not violations:
        return ""

    lines: list[str] = []
    for v in violations:
        level = "error" if v.severity == Severity.ERROR else "warning"
        props: list[str] = []
        if v.location.file is not None:
            props.append(f"file={_escape_gh_property(str(v.location.file))}")
            if v.location.line is not None:
                props.append(f"line={v.location.line}")
        props.append(f"title={_escape_gh_property(v.rule_id)}")
        props_str = ",".join(props)
        lines.append(f"::{level} {props_str}::{_escape_gh_message(v.message)}")

    return "\n".join(lines) + "\n"

File: packages/typescript/src/lint/output.ts
/**
 * Output formatters for lint violations.
 *
 * Three formats:
 *   - `formatHuman`  — pretty terminal output for humans
 *   - `formatJson`   — machine-readable JSON for CI and tooling
 *   - `formatGithub` — GitHub Actions workflow-command annotations
 *
 * The JSON and GitHub formats must produce byte-identical output to
 * the Python implementation given identical inputs.
 */

import type { Severity, Violation } from "./violation.js";

function counts(violations: readonly Violation[]): {
  errors: number;
  warnings: number;
} {
  let errors = 0;
  let warnings = 0;
  for (const v of violations) {
    if (v.severity === "error") errors++;
    else if (v.severity === "warning") warnings++;
  }
  return { errors, warnings };
}

function plural(n: number, word: string): string {
  return n === 1 ? word : `${word}s`;
}

// ---------------------------------------------------------------- human

/**
 * Format violations as human-readable terminal output.
 *
 * @param violations - The lint violations to format.
 * @returns Multi-line string with file locations, messages, and a summary count.
 */
export function formatHuman(violations: readonly Violation[]): string {
  if (violations.length === 0) return "No violations found.\n";

  const lines: string[] = [];
  for (const v of violations) {
    const loc = v.location;
    let prefix: string;
    if (loc.file !== null && loc.line !== null) prefix = `${loc.file}:${loc.line}`;
    else if (loc.file !== null) prefix = loc.file;
    else prefix = "<unknown>";

    lines.push(`${prefix}: ${v.severity}[${v.ruleId}]`);
    lines.push(`  ${v.message}`);
    lines.push(`  Symbolic path: ${loc.symbolic}`);
    if (v.hint) lines.push(`  hint: ${v.hint}`);
    lines.push("");
  }

  const { errors, warnings } = counts(violations);
  const total = errors + warnings;
  const summary =
    `Found ${total} ${plural(total, "violation")} ` +
    `(${errors} ${plural(errors, "error")}, ` +
    `${warnings} ${plural(warnings, "warning")}).`;
  lines.push(summary);
  return lines.join("\n") + "\n";
}

// ---------------------------------------------------------------- json

interface JsonViolation {
  rule_id: string;
  severity: Severity;
  message: string;
  location: { file: string | null; line: number | null; symbolic: string };
  hint: string | null;
}

function violationToJson(v: Violation): JsonViolation {
  return {
    rule_id: v.ruleId,
    severity: v.severity,
    message: v.message,
    location: {
      file: v.location.file,
      line: v.location.line,
      symbolic: v.location.symbolic,
    },
    hint: v.hint ?? null,
  };
}

/**
 * Format violations as a machine-readable JSON string.
 *
 * The output schema matches the Python implementation for cross-language
 * compatibility.
 *
 * @param violations - The lint violations to format.
 * @returns Pretty-printed JSON with `violations` array and `summary` counts.
 */
export function formatJson(violations: readonly Violation[]): string {
  const { errors, warnings } = counts(violations);
  const payload = {
    violations: violations.map(violationToJson),
    summary: { errors, warnings },
  };
  return JSON.stringify(payload, null, 2);
}

// ---------------------------------------------------------------- github

function escapeMessage(msg: string): string {
  return msg
    .split("%").join("%25")
    .split("\r").join("%0D")
    .split("\n").join("%0A");
}

function escapeProperty(value: string): string {
  return value
    .split("%").join("%25")
    .split("\r").join("%0D")
    .split("\n").join("%0A")
    .split(":").join("%3A")
    .split(",").join("%2C");
}

/**
 * Format violations as GitHub Actions workflow-command annotations.
 *
 * Each violation becomes a `::error` or `::warning` line that GitHub
 * renders as an inline annotation on the PR diff.
 *
 * @param violations - The lint violations to format.
 * @returns Newline-delimited workflow commands, or `""` if there are no violations.
 */
export function formatGithub(violations: readonly Violation[]): string {
  if (violations.length === 0) return "";

  const lines: string[] = [];
  for (const v of violations) {
    const level = v.severity === "error" ? "error" : "warning";
    const props: string[] = [];
    if (v.location.file !== null) {
      props.push(`file=${escapeProperty(v.location.file)}`);
      if (v.location.line !== null) props.push(`line=${v.location.line}`);
    }
    props.push(`title=${escapeProperty(v.ruleId)}`);
    lines.push(`::${level} ${props.join(",")}::${escapeMessage(v.message)}`);
  }

  return lines.join("\n") + "\n";
}
